### PR TITLE
Reconfiguration mechanism improvements

### PR DESCRIPTION
**Pull request by yontyon**
In this PR we introduce several improvements, refactoring and bug fixes in the reconfiguration engine:
Below in comments, we elaborate on every change and its purpose

### CODE REVIEW

File: bftengine/include/bftengine/ControlStateManager.hpp [hunk 1]
```diff
@@ -45,9 +45,6 @@ class ControlStateManager {
   void checkForReplicaReconfigurationAction();
   void restart();
 
-  std::pair<bool, std::string> canUnwedge(bool bft);
```
Comment (yontyon) [inline bftengine/include/bftengine/ControlStateManager.hpp:48]: refactoring

File: bftengine/src/bcstatetransfer/BCStateTran.cpp [hunk 2]
```diff
@@ -323,6 +323,14 @@ void BCStateTran::init(uint64_t maxNumOfRequiredStoredCheckpoints,
     LOG_INFO(getLogger(),
              "Init BCStateTran object:" << KVLOG(
                  maxNumOfStoredCheckpoints_, numberOfReservedPages_, config_.sizeOfReservedPage));
+    {
```
Comment (yontyon) [inline bftengine/src/bcstatetransfer/BCStateTran.cpp:326]: Fix a bug where a replica is crashing before full initialization

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 3]
```diff
@@ -182,6 +182,7 @@ IReplica::IReplicaPtr IReplica::createNewReplica(const ReplicaConfig &replicaCon
 
     // Init the persistent storage
     ((PersistentStorageImp *)persistentStoragePtr.get())->init(move(metadataStoragePtr));
+    persistentStoragePtr->setNewEpochFlag(false);
```
Comment (yontyon) [inline bftengine/src/bftengine/BFTEngine.cpp:185]: set to false the nre epoch flag (to handler cases where we don't remove the metadata)

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 4]
```diff
@@ -72,61 +72,6 @@ void ControlStateManager::checkForReplicaReconfigurationAction() {
   }
 }
 
-std::pair<bool, std::string> ControlStateManager::canUnwedge(bool bft) {
```
Comment (yontyon) [inline bftengine/src/bftengine/ControlStateManager.cpp:75]: refactoring

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 5]
```diff
@@ -3298,7 +3298,6 @@ void ReplicaImp::onMessage<ReplicaRestartReadyMsg>(ReplicaRestartReadyMsg *msg)
   if (restart_ready_msgs_.size() == targetNumOfMsgs) {
     LOG_INFO(GL, "Target number = " << targetNumOfMsgs << " of restart ready msgs are recieved. Send resatrt proof");
     sendReplicasRestartReadyProof();
-    ControlStateManager::instance().onRestartProof(msg->seqNum());
```
Comment (yontyon) [inline bftengine/src/bftengine/ReplicaImp.cpp:3301]: As we async sending the messages, I prefer to have the replica handle the restart messages only when it gets it from itself (this way we reduce the probability to restart before completing the actual sending)

File: kvbc/include/reconfiguration_kvbc_handler.hpp [hunk 6]
```diff
@@ -25,9 +25,15 @@ class ReconfigurationBlockTools {
  protected:
   ReconfigurationBlockTools(kvbc::IBlockAdder& block_adder, kvbc::IReader& ro_storage)
       : blocks_adder_{block_adder}, block_metadata_{ro_storage}, ro_storage_{ro_storage} {}
-  kvbc::BlockId persistReconfigurationBlock(const std::vector<uint8_t>& data, const uint64_t bft_seq_num, string key);
+  kvbc::BlockId persistReconfigurationBlock(const std::vector<uint8_t>& data,
```
Comment (yontyon) [inline kvbc/include/reconfiguration_kvbc_handler.hpp:28]: persist wedge block on every wedging command (to have a single mechanism to handle all wedging commands)

File: kvbc/include/st_reconfiguraion_sm.hpp [hunk 7]
```diff
@@ -45,7 +45,17 @@ class StReconfigurationHandler {
   template <typename T>
   void deserializeCmfMessage(T& msg, const std::string& strval);
   template <typename T>
-  bool handlerStoredCommand(const std::string& key, uint64_t current_cp_num);
+  bool handleStoredCommand(const std::string& key, uint64_t current_cp_num);
+
+  template <typename T>
+  bool handleWedgeCommands(const T& cmd,
```
Comment (yontyon) [inline kvbc/include/st_reconfiguraion_sm.hpp:51]: refactoring

File: kvbc/include/st_reconfiguraion_sm.hpp [hunk 8]
```diff
@@ -57,6 +67,7 @@ class StReconfigurationHandler {
   bool handle(const concord::messages::KeyExchangeCommand&, uint64_t, uint64_t, uint64_t) { return true; }
   bool handle(const concord::messages::AddRemoveCommand&, uint64_t, uint64_t, uint64_t) { return true; }
   bool handle(const concord::messages::AddRemoveWithWedgeCommand&, uint64_t, uint64_t, uint64_t);
+  bool handle(const concord::messages::RestartCommand&, uint64_t, uint64_t, uint64_t);
```
Comment (yontyon) [inline kvbc/include/st_reconfiguraion_sm.hpp:70]: add n-f support for restart command

File: kvbc/src/Replica.cpp [hunk 9]
```diff
@@ -166,8 +165,9 @@ void Replica::handleWedgeEvent() {
 
   LOG_INFO(logger,
            "stored wedge info " << KVLOG(wedgePoint, wedgeBlock, wedgeEpoch, lastExecutedSeqNum, latestKnownEpoch));
-  if (wedgeEpoch == latestKnownEpoch && wedgePoint == (uint64_t)lastExecutedSeqNum) {
+  if (wedgeEpoch == latestKnownEpoch && (wedgePoint == (uint64_t)lastExecutedSeqNum || lastExecutedSeqNum == 0)) {
```
Comment (yontyon) [inline kvbc/src/Replica.cpp:168]: handle the case where we still wedged after removing metadata)

File: kvbc/src/Replica.cpp [hunk 10]
```diff
@@ -166,8 +165,9 @@ void Replica::handleWedgeEvent() {
...
     bftEngine::ControlStateManager::instance().setStopAtNextCheckpoint(wedgeBftSeqNum);
+    bftEngine::IControlHandler::instance()->onStableCheckpoint();
```
Comment (yontyon) [inline kvbc/src/Replica.cpp:170]: set the actual wedge point

File: kvbc/src/reconfiguration_kvbc_handler.cpp [hunk 11]
```diff
@@ -336,41 +357,61 @@ bool ReconfigurationHandler::handle(const messages::UnwedgeCommand& cmd,
...
-    ver_updates.addUpdate(std::string{kvbc::keyTypes::reconfiguration_epoch_key},
-                          concordUtils::toBigEndianStringBuffer(newEpoch));
-    auto block_id = persistReconfigurationBlock(ver_updates, bft_seq_num);
-    bftEngine::EpochManager::instance().setSelfEpochNumber(newEpoch);
-    bftEngine::EpochManager::instance().setGlobalEpochNumber(newEpoch);
-    LOG_INFO(getLogger(), "Starting new epoch " << KVLOG(newEpoch, block_id));
-    controlStateManager.setStopAtNextCheckpoint(0);
-    bftEngine::IControlHandler::instance()->resetState();
-    LOG_INFO(getLogger(), "Unwedge command completed sucessfully");
+  auto curr_epoch = bftEngine::EpochManager::instance().getSelfEpochNumber();
```
Comment (yontyon) [inline kvbc/src/reconfiguration_kvbc_handler.cpp:363]: refactoring + change the unwedge signature to contain the replica's epoch number

File: kvbc/src/reconfiguration_kvbc_handler.cpp [hunk 12]
```diff
@@ -336,41 +357,61 @@ bool ReconfigurationHandler::handle(const messages::UnwedgeCommand& cmd,
...
   response.replica_id = bftEngine::ReplicaConfig::instance().replicaId;
-  if (!can_unwedge.first) {
-    response.can_unwedge = false;
-    response.reason = can_unwedge.second;
-    LOG_INFO(getLogger(), "Replica is not ready to unwedge. Reason: " << can_unwedge.second);
-  } else {
-    response.can_unwedge = true;
-    response.signature = std::vector<uint8_t>(can_unwedge.second.begin(), can_unwedge.second.end());
-    LOG_INFO(getLogger(), "Replica is ready to unwedge");
+  if (bftEngine::ControlStateManager::instance().getCheckpointToStopAt().has_value()) {
```
Comment (yontyon) [inline kvbc/src/reconfiguration_kvbc_handler.cpp:398]: same as above

File: kvbc/src/st_reconfiguration_sm.cpp [hunk 13]
```diff
@@ -49,28 +49,28 @@ uint64_t StReconfigurationHandler::getStoredEpochNumber(BlockId bid) {
...
+                                                             current_cp_num);
+  handleStoredCommand<concord::messages::AddRemoveCommand>(std::string{kvbc::keyTypes::reconfiguration_add_remove},
+                                                           current_cp_num);
+  handleStoredCommand<concord::messages::AddRemoveWithWedgeCommand>(
       std::string{kvbc::keyTypes::reconfiguration_add_remove, 0x1}, current_cp_num);
-  handlerStoredCommand<concord::messages::PruneRequest>(std::string{kvbc::keyTypes::reconfiguration_pruning_key, 0x1},
-                                                        current_cp_num);
+  handleStoredCommand<concord::messages::PruneRequest>(std::string{kvbc::keyTypes::reconfiguration_pruning_key, 0x1},
+                                                       current_cp_num);
+  handleStoredCommand<concord::messages::WedgeCommand>(std::string{kvbc::keyTypes::reconfiguration_wedge_key},
```
Comment (yontyon) [inline kvbc/src/st_reconfiguration_sm.cpp:64]: fix a small bug (after state transfer we may unwedge and then handle addRemove command)

File: kvbc/src/st_reconfiguration_sm.cpp [hunk 14]
```diff
@@ -49,28 +49,28 @@ uint64_t StReconfigurationHandler::getStoredEpochNumber(BlockId bid) {
...
-  handlerStoredCommand<concord::messages::KeyExchangeCommand>(std::string{kvbc::keyTypes::reconfiguration_key_exchange},
-                                                              current_cp_num);
-  handlerStoredCommand<concord::messages::AddRemoveCommand>(std::string{kvbc::keyTypes::reconfiguration_add_remove},
-                                                            current_cp_num);
-  handlerStoredCommand<concord::messages::AddRemoveWithWedgeCommand>(
+  handleStoredCommand<concord::messages::InstallCommand>(std::string{kvbc::keyTypes::reconfiguration_install_key},
+                                                         current_cp_num);
+  handleStoredCommand<concord::messages::KeyExchangeCommand>(std::string{kvbc::keyTypes::reconfiguration_key_exchange},
+                                                             current_cp_num);
+  handleStoredCommand<concord::messages::AddRemoveCommand>(std::string{kvbc::keyTypes::reconfiguration_add_remove},
```
Comment (yontyon) [inline kvbc/src/st_reconfiguration_sm.cpp:58]: fix a typo (handler bs handle)

File: kvbc/src/st_reconfiguration_sm.cpp [hunk 15]
```diff
@@ -102,7 +102,8 @@ bool StReconfigurationHandler::handle(const concord::messages::WedgeCommand &,
     return true;
   }
 
-  if (command_epoch < last_known_global_epoch) {
+  if (command_epoch < last_known_global_epoch &&
+      bftEngine::ControlStateManager::instance().getCheckpointToStopAt().has_value()) {
```
Comment (yontyon) [inline kvbc/src/st_reconfiguration_sm.cpp:106]: unwedge only if already wedged

File: reconfiguration/cmf/concord.cmf [hunk 16]
```diff
@@ -123,14 +123,16 @@ Msg UnwedgeStatusRequest 32{
...
     bool can_unwedge
+    uint64 curr_epoch
     bytes signature
     string reason
 }
 
 Msg UnwedgeCommand 34{
     uint64 sender
-    list kvpair uint64 bytes signatures
+    list kvpair uint64 UnwedgeStatusResponse unwedges
```
Comment (yontyon) [inline reconfiguration/cmf/concord.cmf:133]: make the unwedge command containing the epoch numbers of the replicas

File: reconfiguration/include/reconfiguration/reconfiguration_handler.hpp [hunk 17]
```diff
@@ -44,6 +44,9 @@ class ReconfigurationHandler : public BftReconfigurationHandler {
               uint64_t,
               concord::messages::ReconfigurationResponse&) override;
   bool handle(const concord::messages::RestartCommand&, uint64_t, concord::messages::ReconfigurationResponse&) override;
+
+ private:
+  void handleWedgeCommands(bool bft_support, bool remove_metadata, bool restart, bool unwedge);
```
Comment (yontyon) [inline reconfiguration/include/reconfiguration/reconfiguration_handler.hpp:49]: refactoring

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 18]
```diff
@@ -3331,6 +3330,7 @@ void ReplicaImp::sendReplicasRestartReadyProof() {
     }
     restartProofMsg->finalizeMessage();
     sendToAllOtherReplicas(restartProofMsg.get());
+    send(restartProofMsg.get(), config_.replicaId);
```
Comment (nkumar04) [inline bftengine/src/bftengine/ReplicaImp.cpp:3333]: is this a send over loopback socket?

File: kvbc/src/reconfiguration_kvbc_handler.cpp [hunk 19]
```diff
@@ -167,7 +193,7 @@ bool ReconfigurationHandler::handle(const concord::messages::WedgeCommand& comma
   std::vector<uint8_t> serialized_command;
   concord::messages::serialize(serialized_command, command);
   auto blockId = persistReconfigurationBlock(
-      serialized_command, bft_seq_num, std::string{kvbc::keyTypes::reconfiguration_wedge_key});
+      serialized_command, bft_seq_num, std::string{kvbc::keyTypes::reconfiguration_wedge_key}, false);
```
Comment (nkumar04) [inline kvbc/src/reconfiguration_kvbc_handler.cpp:196]: is it better to make include_wedge=false (default)?

File: reconfiguration/src/reconfiguration_handler.cpp [hunk 20]
```diff
@@ -63,50 +63,40 @@ bool ReconfigurationHandler::handle(const concord::messages::AddRemoveWithWedgeC
...
           [=]() { bftEngine::ControlStateManager::instance().sendRestartReadyToAllReplica(); });
-    }
+    if (unwedge)
+      bftEngine::IControlHandler::instance()->addOnStableCheckpointCallBack(
+          [=]() { bftEngine::EpochManager::instance().setNewEpochFlag(true); });
   }
-  return true;
 }
 
 bool ReconfigurationHandler::handle(const concord::messages::RestartCommand& command,
```
Comment (nkumar04) [inline reconfiguration/src/reconfiguration_handler.cpp:94]: concord::messages::RestartCommand with restart=False is same as wedge command. Do we need restart flag for this command? is there a use case where we send RestartCommand with restart=False?

File: kvbc/src/reconfiguration_kvbc_handler.cpp [hunk 21]
```diff
@@ -336,41 +357,61 @@ bool ReconfigurationHandler::handle(const messages::UnwedgeCommand& cmd,
...
+    if (unwedge_stat.curr_epoch < curr_epoch) continue;
+    std::string sig_data = std::to_string(id) + std::to_string(unwedge_stat.curr_epoch);
+    auto& sig = unwedge_stat.signature;
+    std::string signature(sig.begin(), sig.end());
+    bool valid = bftEngine::impl::SigManager::instance()->verifySig(
+        id, sig_data.c_str(), sig_data.size(), signature.data(), signature.size());
+    if (valid) valid_sigs++;
+  }
+  LOG_INFO(getLogger(), "verified " << valid_sigs << " unwedge signatures, required quorum is " << quorum_size);
+  bool can_wedge = (valid_sigs >= quorum_size);
```
Comment (nkumar04) [inline kvbc/src/reconfiguration_kvbc_handler.cpp:378]: nit: rename can_wedge to can_unwedge?

File: kvbc/src/reconfiguration_kvbc_handler.cpp [hunk 22]
```diff
@@ -336,41 +357,61 @@ bool ReconfigurationHandler::handle(const messages::UnwedgeCommand& cmd,
...
-    LOG_INFO(getLogger(), "Replica is not ready to unwedge. Reason: " << can_unwedge.second);
-  } else {
-    response.can_unwedge = true;
-    response.signature = std::vector<uint8_t>(can_unwedge.second.begin(), can_unwedge.second.end());
-    LOG_INFO(getLogger(), "Replica is ready to unwedge");
+  if (bftEngine::ControlStateManager::instance().getCheckpointToStopAt().has_value()) {
+    if ((!req.bft_support && !bftEngine::IControlHandler::instance()->isOnNOutOfNCheckpoint()) ||
+        (req.bft_support && !bftEngine::IControlHandler::instance()->isOnStableCheckpoint())) {
+      response.can_unwedge = false;
+      response.reason = "replica is not at wedge point";
```
Comment (nkumar04) [inline kvbc/src/reconfiguration_kvbc_handler.cpp:402]: should we return from here?

File: reconfiguration/src/reconfiguration_handler.cpp [hunk 23]
```diff
@@ -63,50 +63,40 @@ bool ReconfigurationHandler::handle(const concord::messages::AddRemoveWithWedgeC
...
-    if (command.restart) {
+    if (remove_metadata)
+      bftEngine::IControlHandler::instance()->addOnStableCheckpointCallBack(
+          [=]() { bftEngine::ControlStateManager::instance().markRemoveMetadata(); });
+    if (restart)
       bftEngine::IControlHandler::instance()->addOnSuperStableCheckpointCallBack(
           [=]() { bftEngine::ControlStateManager::instance().sendRestartReadyToAllReplica(); });
-    }
+    if (unwedge)
+      bftEngine::IControlHandler::instance()->addOnStableCheckpointCallBack(
```
Comment (nkumar04) [inline reconfiguration/src/reconfiguration_handler.cpp:89]: addOnSuperStableCheckpointCallBack?

File: reconfiguration/src/reconfiguration_handler.cpp [hunk 24]
```diff
@@ -63,50 +63,40 @@ bool ReconfigurationHandler::handle(const concord::messages::AddRemoveWithWedgeC
...
+          [=]() { bftEngine::EpochManager::instance().setNewEpochFlag(true); });
   } else {
-    bftEngine::IControlHandler::instance()->addOnSuperStableCheckpointCallBack([=]() {
-      bftEngine::ControlStateManager::instance().markRemoveMetadata();
-      bftEngine::EpochManager::instance().setNewEpochFlag(true);
-    });
-    if (command.restart) {
+    if (remove_metadata)
+      bftEngine::IControlHandler::instance()->addOnStableCheckpointCallBack(
```
Comment (nkumar04) [inline reconfiguration/src/reconfiguration_handler.cpp:83]: addOnSuperStableCheckpointCallBack?

File: tests/apollo/util/operator.py [hunk 25]
```diff
@@ -69,11 +69,12 @@ def _construct_reconfiguration_unwedge_status(self, bft):
...
         return self._construct_basic_reconfiguration_request(unwedge_status_cmd)
 
-    def _construct_reconfiguration_unwedge_command(self, signatures, bft):
+    def _construct_reconfiguration_unwedge_command(self, unwedges, bft, restart):
         unwedge_cmd = cmf_msgs.UnwedgeCommand()
         unwedge_cmd.sender = 1000
         unwedge_cmd.bft_support = bft
-        unwedge_cmd.signatures = signatures
+        unwedge_cmd.restart = restart
+        unwedge_cmd.unwedges = unwedges
```
Comment (nkumar04) [inline tests/apollo/util/operator.py:77]: i don't see unwedges option in concord.cmf file. am i missing something?

File: tests/apollo/util/operator.py [hunk 26]
```diff
@@ -157,18 +158,17 @@ async def wedge_status(self, quorum=None, fullWedge=True):
         msg = self._construct_reconfiguration_wedge_status(fullWedge)
         return await self.client.read(msg.serialize(), m_of_n_quorum=quorum, reconfiguration=True)
 
-    async def unwedge(self, bft=False):
-        quorum = None
+    async def unwedge(self, bft=False, restart=False, quorum=None):
```
Comment (nkumar04) [inline tests/apollo/util/operator.py:161]: do we need both bft and quorum in the argument?

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 27]
```diff
@@ -1541,6 +1032,33 @@ async def test_addRemoveStatusError(self, bft_network):
             assert status.response.error_msg == 'key_not_found'
             assert status.success is False
 
+    async def wait_for_full_comm(self, bft_network):
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_reconfiguration.py:1035]: nit: name of this function is not very clear to me? is it full_wedge?

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 28]
```diff
@@ -949,37 +715,39 @@ async def test_restart_command(self, bft_network):
...
         await op.restart("hello", bft=False, restart=False)
         await self.validate_stop_on_wedge_point(bft_network, skvbc, fullWedge=True)
-        previous_last_exec_num = await bft_network.get_metric(0, bft_network, "Gauges", "lastExecutedSeqNum")
         bft_network.stop_all_replicas()
         bft_network.start_all_replicas()
-        for r in bft_network.all_replicas():
-            new_exec_sn = await bft_network.get_metric(r, bft_network, "Gauges", "lastExecutedSeqNum")
-            assert(previous_last_exec_num > new_exec_sn)
+        await self.wait_for_full_comm(bft_network)
+        await op.unwedge(bft=True)
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_reconfiguration.py:730]: should this be op.unwedge(bft=False)?

File: tests/apollo/test_skvbc_reconfiguration.py [hunk 29]
```diff
@@ -949,37 +715,39 @@ async def test_restart_command(self, bft_network):
...
+        await self.validate_epoch_number(bft_network, 1, bft_network.all_replicas())
 
-        # 2. Test with bft
+        # 2. Test without bft and without restart
```
Comment (nkumar04) [inline tests/apollo/test_skvbc_reconfiguration.py:733]: # 2. Test without bft and with restart

File: bftengine/src/bftengine/ControlStateManager.cpp [hunk 30]
```diff
@@ -53,13 +53,16 @@ void ControlStateManager::onRestartProof(const SeqNum& seq_num) {
   // It can happen that some replicas receives a restart proof and yet to reach
   // stable checkpoint. We should not rstart replica in that case since
   // configuration update happens on stable checkpoint.
+  hasRestartProofAtSeqNum_.emplace(seq_num);
   if ((restartBftEnabled_ && IControlHandler::instance()->isOnStableCheckpoint()) ||
       IControlHandler::instance()->isOnNOutOfNCheckpoint()) {
-    for (const auto& kv : onRestartProofCbRegistry_) {
-      kv.second.invokeAll();
+    auto seq_num_to_stop_at = getCheckpointToStopAt();
```
Comment (nkumar04) [inline bftengine/src/bftengine/ControlStateManager.cpp:59]: nit: sequence number check is already done in "ReplicaImp::onMessage<ReplicasRestartReadyProofMsg>"

**Review by nkumar04 — APPROVED** [review]
LGTM!
Thanks!

**Pull request merged**